dialogue-core/src/main/java/com/palantir/dialogue/core/NodeSelectionStrategyChannel.java

```diff
@@ -0,0 +1,191 @@
+    private final TaggedMetricRegistry metrics;
+    private final NodeSelectionStrategySelector strategySelector;
+    private final LimitedChannel delegate;
+
+    public NodeSelectionStrategyChannel(
+            String channelName,
+            Random random,
+            Ticker tick,
+            TaggedMetricRegistry metrics,
+            NodeSelectionStrategySelector strategySelector) {
+        this.channelName = channelName;
+        this.random = random;
+        this.tick = tick;
+        this.metrics = metrics;
+        this.strategySelector = strategySelector;
+        this.nodeChannels = new AtomicReference<>(ImmutableList.of());
+        this.nodeSelectionStrategy = new AtomicReference<>(getUpdatedNodeSelectionStrategy(
+                null, nodeChannels.get(), strategySelector.getCurrentStrategy(), metrics, random, tick, channelName));
+        this.delegate = new SupplierChannel(() -> Preconditions.checkNotNull(
+                        nodeSelectionStrategy.get(), "node selection strategy must not be null")
+                .channel());
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        return delegate.maybeExecute(endpoint, request);
+    }
+
+    public void updateChannels(ImmutableList<LimitedChannel> newChannels) {
+        DialogueNodeSelectionStrategy updatedStrategy = strategySelector.setActiveChannels(newChannels);
+        nodeChannels.set(newChannels);
+        nodeSelectionStrategy.getAndUpdate(previousChannel -> getUpdatedNodeSelectionStrategy(
+                previousChannel.channel(), newChannels, updatedStrategy, metrics, random, tick, channelName));
+    }
+
+    private void updateRequestedStrategies(LimitedChannel channel, List<DialogueNodeSelectionStrategy> strategies) {
+        DialogueNodeSelectionStrategy updatedStrategy = strategySelector.updateChannelStrategy(channel, strategies);
+        nodeSelectionStrategy.getAndUpdate(currentStrategy -> {
+            if (updatedStrategy.equals(currentStrategy.strategy())) {
+                return currentStrategy;
+            }
+            return getUpdatedNodeSelectionStrategy(
+                    currentStrategy.channel(), nodeChannels.get(), updatedStrategy, metrics, random, tick, channelName);
+        });
+    }
+
+    LimitedChannel wrap(LimitedChannel channel) {
+        return new WrapperChannel(channel);
+    }
+
+    private static ChannelWithStrategy getUpdatedNodeSelectionStrategy(
+            @Nullable LimitedChannel previousNodeSelectionStrategy,
+            @Nullable ImmutableList<LimitedChannel> nullableChannels,
+            DialogueNodeSelectionStrategy updatedStrategy,
+            TaggedMetricRegistry metrics,
+            Random random,
+            Ticker tick,
+            String channelName) {
+
+        ImmutableList<LimitedChannel> channels =
+                Preconditions.checkNotNull(nullableChannels, "channels must not be null");
```

Comment:
This doesn't look very nullable

Reply:
You and I both think that, but NullAway insisted that it could be. The alternative was to disable null away on the class, but I went with this so that we would keep the safety if any changes were made

Reply:
Doh. The name and check confused me, I wonder if we could move the null check to call sites instead, if that doesn’t look too messy?